multigrid.tex

```diff
@@ -1 +1,93 @@
-\subsection{\label{sec:multigrid}Multigrid}.
\ No newline at end of file
+\subsection{\label{sec:multigrid}Multigrid}
+The implementation of the multigrid solver algorithm in LFRic has taken a
+long time to develop for several reasons. Firstly, the new solver API,
+which enables easy configuration of different solver algorithms and
+pre-conditioners\footnote{Even nested solvers as currently deployed.},
+uses advanced Fortran OO abstract types. This has caused most
+compilers to have problems and significantly delayed the
+implementation. Secondly, implementing multigrid itself required
+several infrastructure changes. Multiple meshes, maps between these
+meshes and support for intergrid kernels in the PSyKAl API and in
+PSyclone. These have all been successfully implemented. Thirdly, when
+the multigrid algorithm itself was implemented the solver had problems
+converging.
+This was ultimately resolved as an issue with the vertical
+pre-conditioner and not multigrid itself. 
+
+The multigrid algorithm implemented here is a four-level V-cycle of
```

Comment:
It's not really pertinent here but it's bugging me: Why (only) 4 levels?

Reply:
4 is not magic itself, but 4 tends work magically. 3 or less tends not to be enough, 5 tends to be too much, or at least no real gain but one also needs to be have at least 2**5=32 on the finest mesh per parallel MPI rank in order to have a minimum 1 horizontal degree of freedom per MPI rank on the coarsest mesh. Less than 1 requires messy reorganisation. Its usually an empirical measure.

Reply:
Other multigrid cycles are available. W and "full multigrid" which starts on the coarsest and goes up and down 1 level, then 2 and 1, etc. The idea being that spending more algorithm time on the coarsest levels where it is cheap to do the work. There are examples where this has been beneficial but often simplest is best and V-cycle is probably sufficient for us as we only solve to 10e-3 to 10e-6 anyway.

Reply:
Other multi grids are available. For example Algebraic. We are using Geometric multigrid. Our Prolongation and restriction operators use geometry (of the mesh). At a naive level MG just gets a cheaper, less good solution on a coarse mesh which starts the solution guess on the fine mesh. The real magic is the slow-to-converge, high-frequency modes on the fine mesh get aliased to low frequency modes on the coarse mesh (and vice versa) so the solution is much more optimal. Its possible to define Algebraic operators based on some kind of schur complement of the matrix one is trying to invert. This is highly non-trivial to do so most people resort to a library at this point. Hypre is a good one but only works with MPI decomposition (no shared memory) PETSc uses it, but it doesn't help if we are hybrid with Open MP or even a GPU implementation. Again, simple is probably best for us, especially as we gain 90% of the benefit with 10% of the work. Yes, V-cycle GMG is only 10% to of the work compared to some semi-exotic AMG thing. We do only solve until the time-step integration is stable, not to some highly accurate pressure field solution. Clear as mud?

Reply:
Thanks for the extra details. I wrote a V-cycle MG solver back in my PhD days and also found that the optimum number of levels to use depended on the local size of the problem. I also got myself into trouble by asking my solver to solve too many levels - from 512**2 ending up with a 4x4 grid (yes we were all young a stupid once upon a time)- and it ruined the performance. In my experience having a minimum grid size - either explicit or implicit as we have here- is sensible.